File: tic_tac_lua/board.lua
local Utils = require("utils")

Board = {size = 3}

local function create_cells(size)
    local cells = {}
    for i = 1, size do
        cells[i] = {}
        for j = 1, size do
            cells[i][j] = ""
        end
    end
    return cells
end

function Board:init()
    local o = {}
    setmetatable(o, self)
    self.__index = self
    self.cells = create_cells(self.size)

    return o
end

function Board:is_valid_position(row, col, full_check)
    if row < 1 or col < 1 or row > self.size or col > self.size then
        return false
    end

    return not full_check or self.cells[row][col] == ""
end

function Board:get_cell(row, col)
    if not self:is_valid_position(row, col, false) then
        error("Invalid row or column value. Try again")
    end

    return self.cells[row][col]
end

function Board:set_cell(row, col, mark)
    if not self:is_valid_position(row, col, true) then
        error("Invalid row or column value. Try again")
    end

    self.cells[row][col] = mark
end

function Board:display()
    print("\n     c1  c2  c3")
    print("   -------------")

    for i = 1, self.size do
        io.write(Utils.f"r{i} |")
        for j = 1, self.size do
            local cell = self.cells[i][j]
            io.write(" " .. cell .. string.rep(" ", 1 - #cell) .. " |")
        end

        print("\n   -------------")
    end

    print("")
end

function Board:is_full()
    for _, row in ipairs(self.cells) do
        for _, cell in ipairs(row) do
            if cell == "" then
                return false
            end
        end
    end

    return true
end

function Board:has_horizontal_win()
    for _, row in ipairs(self.cells) do
        if row[1] ~= "" and row[1] == row[2] and row[2] == row[3] then
            return true
        end
    end

    return false
end

function Board:has_vertical_win()
    for col = 1, self.size do
        if self.cells[1][col] ~= "" and self.cells[1][col] == self.cells[2][col] and self.cells[2][col] == self.cells[3][col] then
            return true
        end
    end

    return false
end

function Board:has_diagonal_win()
    if (self.cells[1][1] ~= "" and self.cells[1][1] == self.cells[2][2] and self.cells[2][2] == self.cells[3][3]) or
       (self.cells[1][3] ~= "" and  self.cells[1][3] == self.cells[2][2] and self.cells[2][2] == self.cells[3][1]) then
        return true
    end
    return false
end

function Board:check_winner()
    return self:has_horizontal_win() or self:has_vertical_win() or self:has_diagonal_win()
end

function Board:set_remaining_cells()
    for i = 1, self.size do
        for j = 1, self.size do
            if self.cells[i][j] == "" then
                self:set_cell(i, j, Utils.Mark.DEF)
            end
        end
    end
end

return Board

File: tic_tac_zig/board.zig
const std = @import("std");
const Utils = @import("utils.zig");

pub const Board = struct {
    size: usize = 3,
    cells: [3][3][]const u8 = undefined,

    pub fn init() Board {
        var matrix = [3][3][]const u8 {
            [_][]const u8{ "", "", "" },
            [_][]const u8{ "", "", "" },
            [_][]const u8{ "", "", "" },
        };

        return Board {
            .cells = matrix,
        };
    }

    pub fn isValidPosition(self: Board, row: usize, col: usize, fullCheck: bool) !bool {
        if (row < 0 or col < 0 or row >= self.size or col >= self.size) {
            return false;
        }

        return !fullCheck or std.mem.eql(u8, self.cells[row][col], Utils.emptyString);
    }

    pub fn getCell(self: Board, row: usize, col: usize) ![]u8 {
        if (!try self.isValidPosition(row, col, false)) {
            @panic("Invalid row or column value. Try again");
        }

        return self.cells[row][col];
    }

    pub fn setCell(self: Board, row: usize, col: usize, mark: []const u8) !void {
        if (!try self.isValidPosition(row, col, true)) {
            @panic("Invalid row or column value. Try again");
        }
        self.cells[row][col] = mark;
    }

    pub fn display(self: Board) !void {
        const out = std.io.getStdOut().writer();


        try out.print("\n     c1  c2  c3\n", .{});
        try out.print("   -------------\n", .{});

        for (0..self.size) |i| {
            try out.print("r{d} |", .{i + 1});
            for (0..self.size) |j| {
                if (std.mem.eql(u8, self.cells[i][j], Utils.emptyString)) {
                    try out.print("  {s} |", .{self.cells[i][j]});
                    
                } else {
                    try out.print(" {s} |", .{self.cells[i][j]});
                }
            }
            try out.print("\n   -------------\n", .{});
        }

        try out.print("\n", .{});
    }

    pub fn isFull(self: Board) bool {
        var full = true;
        for (self.cells) |row| {
            for (row) |cell| {
                if (std.mem.eql(u8, cell, "")) {
                    full = false;
                    break;
                }
            }
            if (!full) {
                break;
            }
        }
        return full;
    }

    pub fn hasHorizontalWin(self: Board) bool {
        for (0..self.size) |i| {
            var allEqual = true;
            const firstEl = self.cells[i][0];
            for (self.cells[i]) |cell| {
                if (!std.mem.eql(u8, cell, firstEl) or std.mem.eql(u8, cell, "")) {
                    allEqual = false;
                    break;
                }
            }

            if (allEqual) {
                return true;
            }
        }

        return false;
    }

    pub fn hasVerticalWin(self: Board) bool {
        var columns: [self.size][][]const u8 = undefined;

        for (0..self.size) |i| {
             for (0..self.size) |j| {
                columns[j][i] = self.cells[i][j];
             }
        }

        for (columns) |col| {
            var allEqual = true;
            const firstEl = col[0];
            for (col) |cell| {
                if (!std.mem.eql(u8, cell, firstEl) or std.mem.eql(u8, cell, "")) {
                    allEqual = false;
                    break;
                }
            }

            if (allEqual) {
                return true;
            }
        }
        return false;
    }

    pub fn hasDiagonalWin(self: Board) bool {
        const mainDiag: [self.size][]const u8 = undefined;
        const secDiag: [self.size][]const u8 = undefined;

        for (0..self.size) |i| {
            mainDiag[i] = self.cells[i][i];
            secDiag[i] = self.cells[i][self.size - i - 1];
        }

        var mainMatch: bool = !std.mem.eql(u8, mainDiag[0], "");

        if (mainMatch) {
            for (mainDiag) |cell| {
                if (std.mem.eql(u8, cell, mainDiag[0])) {
                    mainMatch = true;
                    break;
                }
            }
        }

        var secMatch: bool = !std.mem.eql(u8, secDiag[0], "");
        if (secMatch) {
            for (secDiag) |cell| {
                if (std.mem.eql(u8, cell, secDiag[0])) {
                    secMatch = true;
                    break;
                }
            }
        }

        return mainMatch or secMatch;
    }

    pub fn checkWinner(self: Board) bool {
        return self.hasHorizontalWin() or self.hasVerticalWin() or self.hasDiagonalWin();
    }

    pub fn setRemainingCells(self: Board) void {
        for (0..self.size) |i| {
            for (0..self.size) |j| {
                if (std.mem.eql(u8, self.cells[i][j], "")) {
                    self.cells[i][j] = Utils.Mark.DEF;
                }
            }
        }
    }
};